### PR TITLE
[CURATOR-223] Add executorService methods to ServiceCacheBuilder

### PR DESCRIPTION
**Pull request by tdyas**
Add .executorService variants to ServiceCacheBuilder to allow the caller to specify an ExecutorService or a CloseableExecutorService to be used by the PathChildrenCache embedded in ServiceCacheImpl.

Extracts ExecuteCalledWatchingExecutorService (and DelegatingExecutorService) into the curator-test module for use by TestServiceCache.

Relates to CURATOR-223

### CODE REVIEW

File: curator-recipes/src/main/java/org/apache/curator/framework/recipes/cache/PathChildrenCache.java [hunk 1]
```diff
@@ -184,6 +184,28 @@ public PathChildrenCache(CuratorFramework client, String path, boolean cacheData
     }
 
     /**
+     * @param client        the client
```
Comment (Randgalt) [inline curator-recipes/src/main/java/org/apache/curator/framework/recipes/cache/PathChildrenCache.java]: Why are these new ctors need? PCC already has a bazillion

File: curator-x-discovery/src/main/java/org/apache/curator/x/discovery/details/ServiceCacheImpl.java [hunk 2]
```diff
@@ -66,6 +68,30 @@
         cache.getListenable().addListener(this);
     }
 
+    ServiceCacheImpl(ServiceDiscoveryImpl<T> discovery, String name, ExecutorService executorService)
```
Comment (Randgalt) [inline curator-x-discovery/src/main/java/org/apache/curator/x/discovery/details/ServiceCacheImpl.java]: I'd prefer to see 1 ctor. It can take a CloseableExecutorService. The builder can wrap the executorService. I think there are other examples of this in the Curator code.

Comment (tdyas) [inline curator-x-discovery/src/main/java/org/apache/curator/x/discovery/details/ServiceCacheImpl.java]: Makes sense. Next iteration will have one ctor.

File: curator-recipes/src/main/java/org/apache/curator/framework/recipes/cache/PathChildrenCache.java (hunk 1 above)
Comment (tdyas) [inline curator-recipes/src/main/java/org/apache/curator/framework/recipes/cache/PathChildrenCache.java]: I was trying to avoid using the ctor's with the `dataIsCompressed` argument since `ServiceCacheImpl` did not use those ctor's. I suppose it is fine to just invoke those ctor's and pass in `false` for `dataIsCompressed`.

**Comment by tdyas** [comment]
I made the requested updates. The proposed commit no longer modifies `PathChildrenCache`, and `ServiceCacheImpl` has one additional ctor.

File: curator-x-discovery/src/main/java/org/apache/curator/x/discovery/details/ServiceCacheImpl.java [hunk 3]
```diff
@@ -66,6 +68,18 @@
         cache.getListenable().addListener(this);
     }
 
+    ServiceCacheImpl(ServiceDiscoveryImpl<T> discovery, String name, CloseableExecutorService executorService)
```
Comment (Randgalt) [inline curator-x-discovery/src/main/java/org/apache/curator/x/discovery/details/ServiceCacheImpl.java]: Please have the other ctor call this one. You may need to introduce a third private ctor. i.e. only 1 ctor should have code. The others should forward to the 1 with code.

File: curator-recipes/src/test/java/org/apache/curator/framework/recipes/cache/TestPathChildrenCache.java [hunk 4]
```diff
@@ -1039,127 +1040,4 @@ public void invoke() throws Exception
             CloseableUtils.closeQuietly(client);
         }
     }
-
-    public static class ExecuteCalledWatchingExecutorService extends DelegatingExecutorService
```
Comment (Randgalt) [inline curator-recipes/src/test/java/org/apache/curator/framework/recipes/cache/TestPathChildrenCache.java]: Why were these moved to separate files?

Comment (tdyas) [inline curator-recipes/src/test/java/org/apache/curator/framework/recipes/cache/TestPathChildrenCache.java]: To allow reuse in `TestServiceCache`.

**Comment by Randgalt** [comment]
Please rename this PR so that the Jira Issue number is in the subject (see the other PRs)

File: curator-x-discovery/src/main/java/org/apache/curator/x/discovery/details/ServiceCacheImpl.java (hunk 3 above)
Comment (tdyas) [inline curator-x-discovery/src/main/java/org/apache/curator/x/discovery/details/ServiceCacheImpl.java]: Will do. I had to add a conversion from `ThreadFactory` to `CloseableExecutorService` in a static method since Java doesn't permit ctor's to have code before the call to the common ctor.
